### PR TITLE
RFC: Node.js Installation Buildpacks BOM

### PR DESCRIPTION
**Pull request by sophiewigmore**
## Summary

Resolves https://github.com/paketo-buildpacks/nodejs/issues/425

[Readable](https://github.com/paketo-buildpacks/rfcs/blob/install-bom/text/nodejs/0013-install-bom.md)

## Checklist

* [x] I have viewed, signed, and submitted the Contributor License Agreement.
* [x] I have linked issue(s) that this PR should close using keywords or the Github UI (See [docs](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue))
* [ ] I have added an integration test, if necessary.
* [x] I have reviewed the [styleguide](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md) for guidance on my code quality.
* [ ] I'm happy with the commit history on this PR (I have rebased/squashed as needed).

### CODE REVIEW

File: text/nodejs/0013-install-bom.md
Comment (arjun024) [inline text/nodejs/0013-install-bom.md:70]: NPM Install/Yarn Install buildpacks are designed such that they can be used offline in Internet-less environments. This installation of cyclonedx/bom doesn't seem to work in offline mode. Can this be packaged as a dependency? Or even, do you think this merits a buildpack of its own or be in the build image? Regardless, I think we should maintain install buildpack's ability to work fully in offline mode.

Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md:70]: @arjun024 that's a good point. I'm more inclined to have this run as a part of this buildpack since the goal is just to collect insights around the install process, and it doesn't really have a role in adding something to the final app image (besides metadata). 

I can look into packaging this as a dependency and update you here.

Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md:70]: I think @ForestEckhardt and I have two options we are considering after looking into this:

A. Extract the tool as a tarball with its dependencies, and make it available as a standalone dep-server dependency. Install it with it’s own buildpack in both an online and offline scenario. Run BOM creation using the tool in its own buildpack after NPM Install/Yarn Install/ or before Node Start in the no package manager scenario. 

B. Extract the tool as a tarball with its dependencies, and then run BOM tool installation and BOM creation together in one buildpack after NPM Install/Yarn Install/ before Node Start in the no package manager scenario.

We are leaning towards having this process be a part of its own buildpack(s) for a couple reasons. First, it keeps buildpack modularity intact. While it seems intuitive to run BOM creation alongside the NPM Install and Yarn Install buildpacks, the issue is that we need to support the Node Start case, and it seems unintuitive to create the BOM in this buildpack, since it's a start-command buildpack.
 @arjun024 do you have any thoughts?
The other reason we are leaning to one of these two set ups is because it feels slightly sneaky to slip the dependency into the image within another buildpack. We could add it as a part of the node-engine dependency or install it in the node-engine buildpack, but this doesn't seem that transparent especially since the inclusion of this dependency opens the door for additional CVEs.

Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md:70]: We are ideally going with option B here. We will host the CycloneDX BOM tool as a dependency in the dep-server, and then introduce a buildpack that runs before start commands to install the tool and run BOM generation together.

Comment (arjun024) [reply text/nodejs/0013-install-bom.md:70]: SGTM

File: text/nodejs/0013-install-bom.md
Comment (arjun024) [inline text/nodejs/0013-install-bom.md:76]: Nitpick: You are not writing to the "layer metadata" right? You're gonna write to the {build,launch}.toml? (build/launch environment metadata - I don't know the technical term?) I think of layer metadata as `<layers>/<layer>.toml`. I think spelling it out would be clearer.

File: text/nodejs/0013-install-bom.md
Comment (arjun024) [inline text/nodejs/0013-install-bom.md:93]: I don't think apps would have a `node_modules` dir at detect time unless they are vendored. The x-install buildpacks generate node_modules dir during the build phase

File: text/nodejs/0013-install-bom.md
Comment (arjun024) [inline text/nodejs/0013-install-bom.md:106]: >  It will get added to a build-time layer

Why is this tool added to a layer (& a BOM entry for the tool itself added) instead of just using the tool to generate the BOM and throwing away the tool?

File: text/nodejs/0013-install-bom.md
Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md:93]: Yea the point here was that. If the modules are vendored, we don't have to require `node_modules`. My wording is a bit misleading I can fix it up

File: text/nodejs/0013-install-bom.md
Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md:106]: We wanted to cache the tool so that if any node modules change, we can still generate the BOM again without having to re-download the tool.

A BOM entry for the tool is necessary because it's something we are installing onto the image, and we need to record of everything we install via our buildpacks, same as we do for node engine and. yarn

File: text/nodejs/0013-install-bom.md
Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md]: tried to fix the wording in https://github.com/paketo-buildpacks/rfcs/pull/98/commits/7f6e30ec1eedcbb1221f9472bb97fa3464887a54

File: text/nodejs/0013-install-bom.md
Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md]: See https://github.com/paketo-buildpacks/rfcs/pull/98/commits/7f6e30ec1eedcbb1221f9472bb97fa3464887a54

File: text/nodejs/0013-install-bom.md
Comment (arjun024) [reply text/nodejs/0013-install-bom.md]: Can we fail detection if BP_*_BOM is set to false? - so we can use it as an optional buildpack. If that's too detailed for the RFC, that's fine too.

File: text/nodejs/0013-install-bom.md
Comment (arjun024) [inline text/nodejs/0013-install-bom.md:86]: Are you planning to use this as a generic flag across language families? If it's nodejs-specific, I suggest we stick to the convention of the BP_NODE prefix

File: text/nodejs/0013-install-bom.md
Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md]: Yea!! That's what we're proposing. Is that unclear in our explanation?

File: text/nodejs/0013-install-bom.md
Comment (ryanmoran) [reply text/nodejs/0013-install-bom.md:86]: It may make sense to have both. Imagine that a user might want to disable the BOM for just a single buildpack, or all BOM generation across the entire build.

Comment (ForestEckhardt) [reply text/nodejs/0013-install-bom.md:86]: How would y'all feel if for this RFC and the initial implementation of the buildpack we removed the ability to skip detection via and environment variable. I think that we could hold off for now and survey the landscape of language module BOM generating buildpacks and make a more informed decision on what kind of control mechanism we would like. I think that this is a relatively safe approach because if we get user feedback in the interim for such a feature this control structure is easily added and will help us solidify a pattern for similar types of buildpacks that may come in the future.

Comment (arjun024) [reply text/nodejs/0013-install-bom.md:86]: If the bom generation do not eat up a lot time, I'm fine with it. You can also propose the skip-detection in the RFC and not do it in the initial implementation if you want.

Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md]: @arjun024 we have updated the RFC to mention this environment variable as a future optimization pending user interest. We are leaving it out of our initial implementation

File: text/nodejs/0013-install-bom.md
Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md]: Since we are no longer implementing this environment variable, I don't think this is necessary now?

**Review by arjun024 — APPROVED** [review]

File: text/nodejs/0013-install-bom.md
Comment (ekcasey) [inline text/nodejs/0013-install-bom.md:78]: I wonder if instead of generating the BOM during the build phase of the BOM generation buildpack, we should instead have the buildpack that produces installs the modules generate the BOM. There are a couple reasons for this:

Once CNB introduces the ability [associate BOM entries with a specific layer](https://github.com/buildpacks/rfcs/blob/main/text/0087-bom-in-layer-metadata.md), we will want to:
1. e.g. associate these BOM entries with a layer like paketo-buildpacks/yarn-install:modules
2. save time by avoiding rerunning the tool if the layer has not changed (the yarn-install buildpack should be responsible for making this determination)

In the case where there is no package manage we will still need the new buildpack to generate the BOM, or I guess node-start could potentially do it?

File: text/nodejs/0013-install-bom.md
Comment (ekcasey) [inline text/nodejs/0013-install-bom.md:88]: Seems like it should provide cyclonedx-bom. If we take my sugestion above regarding, shifting responsibility for executing the tool to the buildpack providing the described layer (when appropriate), then
1. if `node_modules` is vendored, it will also require cyclonedx-bom
2. yarn-install and npm-install can require cyclonedx-bom in their first build plan entry (and fall back to passing without generating a bom in a second entry).

File: text/nodejs/0013-install-bom.md
Comment (ForestEckhardt) [reply text/nodejs/0013-install-bom.md:78]: I think that we need a new buildpack regardless and as it stands we can get away with having one buildpack do all of the BOM generation. I would prefer to implement the smallest change first and then potentially expand the functionality of other buildpacks or add additional buildpacks as the need arises. I have some concerns about having to duplicate logic across three different buildpacks just so I have the BOM on the right layer and so I would like to dive into this deeper when the feature is ready to play.

Is the lack of consideration for layer specific BOM a blocker for this initial implementation in your eyes?

File: text/nodejs/0013-install-bom.md
Comment (ForestEckhardt) [reply text/nodejs/0013-install-bom.md:88]: I think this is correct if we do divide up the generation amongst buildpacks, however I am hesitant to do this fo the reasons mentioned above.

File: text/nodejs/0013-install-bom.md
Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md:88]: We just added a performance section, but when we checked the time it took to run the tool against an app with 1300 node modules, it only took around 3 seconds. It's pretty quick, so we're not overly worried about the re-running of the tool

Comment (ekcasey) [reply text/nodejs/0013-install-bom.md:88]: > it only took around 3 seconds. It's pretty quick, so we're not overly worried about the re-running of the tool

While I don't think 3 seconds is a deal-breaker, it's not completely insignificant either. For example a no-op rebuild of the paketo yarn sample app currently takes about 4.6 seconds on my workstation.

>  I have some concerns about having to duplicate logic across three different buildpacks just so I have the BOM on the right layer and so I would like to dive into this deeper when the feature is ready to play.

My assumption would be that the vast majority of the logic lives in the bom generation tool itself, leaving very little logic that would need to be duplicated?

> Is the lack of consideration for layer specific BOM a blocker for this initial implementation in your eyes?

No. In general I don't see my suggestions here as a blocker. I think associating the BOM with the correct layer and shaving a few seconds off the build whenever possible are both good long term goals, but if the node maintainers believe that accomplishing those goals would significantly slow down our ability to deliver ther feature or make the buildpacks harder to maintain, I trust the maintainers weigh those tradeoffs. This also seems like something that could be refactored later if need be without impacting end users, which lowers the stakes.

Comment (sophiewigmore) [reply text/nodejs/0013-install-bom.md:88]: > While I don't think 3 seconds is a deal-breaker, it's not completely insignificant either. For example a no-op rebuild of the paketo yarn sample app currently takes about 4.6 seconds on my workstation.

This is fair. The app we used is quite large so I'd expect that to be more of a worst-case time. Regardless, you are right though that it's significant enough that we should probably consider optimizing.
 
> My assumption would be that the vast majority of the logic lives in the bom generation tool itself, leaving very little logic that would need to be duplicated?

Right now the logic includes the installation of the tool (/cache reuse of the tool), running the tool, and converting the BOM JSON file into `packit.BOMEntry` types (for now). 

It's not terrible to duplicate, but the fact of the matter is that we can get the basic functionality for the BOM via a single buildpack, rather than adding a separate buildpack for one case, and making changes to NPM Install and Yarn Install in the other cases. The optimizations for build time and layer inclusion are definitely worthwhile, but I am more inclined to go with the basic implementation that we have outlined first (pending maintainer approval) and then work towards these optimizations down the line as you said.

@paketo-buildpacks/nodejs-maintainers can you weigh in?

Comment (arjun024) [reply text/nodejs/0013-install-bom.md:88]: SGTM for the first implementation. If the `node_modules` hasn't changed, you should still be able to save time by avoiding downloading/rerunning the tool, right?

Comment (ForestEckhardt) [reply text/nodejs/0013-install-bom.md:88]: Correct

Comment (thitch97) [reply text/nodejs/0013-install-bom.md:88]: I'm also on board with the single-buildpack proposal as a first pass

**Review by arjun024 — APPROVED** [review]

**Pull request merged**
